### PR TITLE
Fix indexed access on xml union type

### LINKED ISSUE #35008: Some user defined subtypes of `xml` do not support member access
**Description:**
$title

**Steps to reproduce:**
```ballerina
type XmlType1 xml:Text|xml:Comment;

type XmlType2 xml<xml:Element|xml:ProcessingInstruction>;

type XmlType3 XmlType1|XmlType2;

function init() {
    XmlType1 txt = xml `Hello!`;

    io:println(txt[0]); // compilation error

    io:println(txt[0]); // compilation error

    io:println(txt[0]); // compilation error

    XmlType3 ele2 = xml `<name>Sherlock Holmes</name>`;
    io:println(ele2[0]); // compilation error

    xml:Text|xml:ProcessingInstruction txt2 = xml `World`;
    io:println(txt2[0]); // compilation error

    xml|xml<xml:Text> authors = xml `<name>Sherlock Holmes</name><name>Enid Blyton</name><name>Dan Brown</name><name>Nicholas Sparks</name>`;
    io:println(authors[0]); // compilation error

    xml<xml:Text|xml:Comment>|xml<xml:ProcessingInstruction|xml:Element> ele3 = xml `<name>Dan Brown</name>`;
    io:println(ele3[0]); // compilation error

    XmlType2 ele = xml `<name>Sherlock Holmes</name>`;
    io:println(ele[0]); // works
    
    xml<xml:Element|xml:ProcessingInstruction> ele4 = xml `<name>Enid Blyton</name>`;
    io:println(ele4[0]); // works
}
```

**Affected Versions:**
Ballerina 2201.0.0 (Swan Lake)

### PR DESCRIPTION
**Pull request by poorna2152**
## Purpose
Fixes #35008 

## Approach
Use `isAssignable` check to allow xml union types instead of checking using the xml type tag which does not cover union types.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml-indexed-access.bal [hunk 1]
```diff
@@ -444,6 +444,94 @@ function testXmlAccessWithLargerIndex() {
...
+type XmlType1 xml:Text|xml:Comment;
+
+type XmlType2 xml<xml:Element|xml:ProcessingInstruction>;
+
+type XmlType3 XmlType1|XmlType2;
+
+function testXmlIndexedAccessWithUnionType() {
+    xml:Element|xml:Comment x1 = xml `<a>abc</a>`;
+    assert(x1[0], xml `<a>abc</a>`);
+    assert(x1[1], xml ``);
```
Comment (SasinduDilshara) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml-indexed-access.bal:456]:
```suggestion
    assert(x1[1], xml ``);
    
```

Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml-indexed-access.bal:456]: Imo, it is better to add a new line after each test group to make it clear

File: tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml-indexed-access.bal [hunk 2]
```diff
@@ -444,6 +444,94 @@ function testXmlAccessWithLargerIndex() {
...
+    xml|xml x9 = xml `<c>ghi</c><d>jkl</d>`;
+    assert(x9[0], xml `<c>ghi</c>`);
+    xml<xml:Text|xml:Comment>|xml<xml:ProcessingInstruction|xml:Element> x10 = xml `<name>Dan Brown</name>`;
+    assert(x10[0], xml `<name>Dan Brown</name>`);
+    assert(x10[50], xml ``);
+    XmlType1 x11 = xml `Hello!`;
+    assert(x11[0], xml `Hello!`);
+    XmlType3 x12 = xml `<name>Sherlock Holmes</name>`;
+    assert(x12[0], xml `<name>Sherlock Holmes</name>`);
+    assert(x12[50][1], xml ``);
```
Comment (SasinduDilshara) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml-indexed-access.bal:483]: Shall we add test with `xml:sequence` type as a member of a union type

File: tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml-indexed-access.bal [hunk 3]
```diff
@@ -444,6 +444,94 @@ function testXmlAccessWithLargerIndex() {
...
+type XmlType2 xml<xml:Element|xml:ProcessingInstruction>;
+
+type XmlType3 XmlType1|XmlType2;
+
+function testXmlIndexedAccessWithUnionType() {
+    xml:Element|xml:Comment x1 = xml `<a>abc</a>`;
+    assert(x1[0], xml `<a>abc</a>`);
+    assert(x1[1], xml ``);
+    xml:Element|xml:Text x2 = xml `Hello World`;
+    assert(x2[0], xml `Hello World`);
```
Comment (LakshanWeerasinghe) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/xml/xml-indexed-access.bal:458]: Shall we assert the type as well?
```suggestion
    assert(x2[0], xml `Hello World`);
    assert(x2[0], xml:Text);
```

**Review by KavinduZoysa — APPROVED** [review]

**Pull request merged**